03-pages > 01-building-your-application > 01-routing > 07-api-routes.mdx (#279)
Nextjs-kr/Nextjs.kr: docs/03-pages/01-building-your-application/01-routing/07-api-routes.mdx

export default function handler(req, res) {
  res.status(200).json({ name: 'John Doe' })
}
```

> **참고**: API 라우트는 `next.config.js`에 있는 [`pageExtensions` 설정](/docs/pages/api-reference/next-config-js/pageExtensions)에 영향을 받습니다.

API 라우트가 동작하려면 다음 매개변수를 받는 함수를 default로 내보내야 합니다. (일명 **요청 핸들러**)

- `req`: [http.IncomingMessage](https://nodejs.org/api/http.html#class-httpincomingmessage) 인스턴스와 [미리 만들어진 미들웨어](/docs/pages/api-reference/functions/next-server)
- `res`: [http.ServerResponse](https://nodejs.org/api/http.html#class-httpserverresponse) 인스턴스와 [헬퍼 함수](/docs/pages/api-reference/functions/next-server)

API 라우트에서 다른 HTTP 메서드를 처리하려면, 요청 핸들러에서 `req.method`를 사용하면 됩니다.

```js
export default function handler(req, res) {
  if (req.method === 'POST') {
    // POST 요청 처리
  } else {
    // 다른 HTTP 메서드 처리
  }
}
```

API 엔드포인트를 가져오려면 이 섹션의 시작 부분에 제공된 예시를 살펴보세요.

## 사용 사례

새 프로젝트의 경우, API 라우트를 사용하여 전체 API를 만들 수 있습니다. 기존 API가 있다면, API 라우트를 통해 API 호출을 전달할 필요가 없습니다. API 라우트에 대한 다른 사용 사례는 다음과 같습니다.

- 외부 서비스의 URL을 마스킹하기 (예시: `https://company.com/secret-url` 대신 `/api/secret` )
- 서버에서 [환경 변수](/docs/pages/building-your-application/configuring/environment-variables)를 안전하게 사용하여 외부 서비스에 접근하기.

## 주의 사항

> - API 라우트는 [CORS 헤더를 지정하지 않습니다.](https://developer.mozilla.org/ko/docs/Web/HTTP/CORS) 이는 기본적으로 **동일 출처 전용**임을 의미합니다.

Reviewer: ```suggestion
- API 라우트는 [CORS 헤더를 지정하지 않습니다](https://developer.mozilla.org/ko/docs/Web/HTTP/CORS). 이는 기본적으로 **동일 출처 전용**임을 의미합니다.
```

}
```

`api` 객체는 API 라우트에 사용할 수 있는 모든 설정 옵션을 포함합니다.

`bodyParser`는 자동으로 활성화됩니다. 본문을 `Stream`으로 소비하거나 [`raw-body`](https://www.npmjs.com/package/raw-body)를 사용하려면 이를 `false`로 설정할 수 있습니다.

> 자동 `bodyParsing`을 비활성화하는 사용 사례는 **웹훅** 요청의 원시 본문을 확인하려는 경우입니다. 예를 들어 [GitHub으로부터의 웹훅](https://docs.github.com/ko/webhooks-and-events/webhooks/securing-your-webhooks) 요청을 받을 수 있습니다.

Reviewer: ```suggestion
자동 `bodyParsing`을 비활성화하는 사용 사례는 **웹훅** 요청의 원시 본문을 확인하려는 경우입니다. 웹훅의 예시로는 [Github 웹훅](https://docs.github.com/ko/webhooks-and-events/webhooks/securing-your-webhooks)이 있습니다.
```

}
```

> `externalResolver`는 이 라우트가 _express_ 또는 *connect*와 같은 외부 리졸버(resolver)에 의해 처리되고 있음을 서버에 명시적으로 알려주는 플래그입니다. 이 옵션을 활성화하면 미해결 요청에 대한 경고가 비활성화됩니다.

Reviewer: ```suggestion
`externalResolver`는 이 라우트가 _express_ 또는 _connect_와 같은 외부 리졸버(resolver)에 의해 처리되고 있음을 서버에 명시적으로 알려주는 플래그입니다. 이 옵션을 활성화하면 미해결 요청에 대한 경고가 비활성화됩니다.
```

}
```

> `responseLimit`은 바이트 수 또는 `bytes`가 지원하는 어떤 문자열 형식도 받을 수 있습니다. 예를 들어 `1000`, `500kb` 또는 `3mb` 등이 가능합니다.

Reviewer: ```suggestion
`responseLimit`은 바이트 수 또는 `bytes`가 지원하는 어떤 문자열 형식도 받을 수 있습니다. 예를 들어 `1000` 또는 `'500kb'`, `'3mb'` 등이 가능합니다.
```

import { withFoo } from 'external-lib-foo'

const handler = (req, res) => {
  req.foo('bar') // 이제 타입 에러 없이 `req.foo`를 사용할 수 있습니다
  res.end('ok')
}

export default withFoo(handler)
```

`withFoo()`를 export에서 제거하더라도 코드는 여전히 컴파일되므로, 이는 안전하지 않다는 점을 유의하세요.

## 응답 헬퍼

[서버 응답 객체](https://nodejs.org/api/http.html#http_class_http_serverresponse)(보통 `res`로 줄여서 표현됨)는 개발자 경험을 향상하고 새로운 API 엔드포인트를 빠르게 생성하는 데 도움을 주기 위해 Express.js와 유사한 여러 헬퍼 함수를 포함하고 있습니다.

포함된 여러 헬퍼는 다음과 같습니다.

> - `res.status(code)` - 상태 코드를 설정하는 함수입니다. `code`는 유효한 [HTTP 상태 코드](https://ko.wikipedia.org/wiki/HTTP_%EC%83%81%ED%83%9C_%EC%BD%94%EB%93%9C)여야 합니다.

Reviewer: 왠만한 외부 레퍼런스 링크에 대해 한글로 번역된 링크로 변환해주셔서 감사합니다. 그런데 해당 링크에 대해 영문 그대로 놔두는 것이 현재 번역 진행 상황이라 우선 커멘트 남깁니다.

- `res.json(body)` - JSON 응답을 보냅니다. `body`는 [직렬화할 수 있는 객체](https://developer.mozilla.org/en-US/docs/Glossary/Serialization)여야 합니다.
- `res.send(body)` - HTTP 응답을 보냅니다. `body`는 `string`, `object` 또는 `Buffer`일 수 있습니다.
- `res.redirect([status,] path)` - 지정된 경로 또는 URL로 리다이렉트합니다. `status`는 유효한 [HTTP 상태 코드](https://ko.wikipedia.org/wiki/HTTP_%EC%83%81%ED%83%9C_%EC%BD%94%EB%93%9C)여야 합니다.
  지정하지 않는다면 `status`의 기본값은 "307" 즉, "임시 리다이렉트(Temporary redirect)"입니다.

> - `res.revalidate(urlPath)` - `getStaticProps`를 사용하여 [요청 시 페이지 재검증](/docs/pages/building-your-application/data-fetching/incremental-static-regeneration#on-demand-revalidation)을 수행합니다. `urlPath`는 `string`여야 합니다.

Reviewer: 'on demand'를 '요청 시'로 번역하셨군요! 번역 잘 해주셨고 문제 없지만 나중에 통일할 필요가 있어 보여 커멘트 남깁니다.

}
```

`/api/post/abc`로의 요청은 텍스트 `Post: abc`를 응답으로 받게 됩니다.

### 인덱스 라우트와 동적 API 라우트

매우 일반적인 RESTful 패턴은 다음과 같이 라우트를 설정하는 것입니다.

- `GET api/posts` - 페이징 처리된 포스트 목록을 가져옵니다.
- `GET api/posts/12345` - 포스트 id 12345를 가져옵니다.

이를 두 가지 방법으로 모델링할 수 있습니다.

- 옵션 1
  - `/api/posts.js`
  - `/api/posts/[postId].js`
- 옵션 2
  - `/api/posts/index.js`
  - `/api/posts/[postId].js`

두 가지 옵션은 동일합니다. 세 번째 옵션인 `/api/posts/[postId].js`만 사용하는 것은 유효하지 않습니다.
그 이유는 동적 라우트(아래의 캐치-올 라우트 포함)는 `undefined` 상태를 갖지 않으며, 어떠한 상황에서도 `GET /api/posts`는 `/api/posts/[postId].js`와 일치하지 않기 때문입니다.

> ### 캐치 올 API 라우트

Reviewer 2: 저는 `Catch-all Segments` => `모든 세그먼트 캡처하기`로 번역했습니다.
`모든 API 경로 캡처하기`는 어떨까요?
브라우저의 캡처링과 관련하여 통일성도 충족할 것 같아요.

}
```

이제 `/api/post/a/b/c`로의 요청은 텍스트 `Post: a, b, c`를 응답으로 받게 됩니다.

### 선택적 캐치 올 API 라우트

캐치 올 라우트는 매개변수를 이중 괄호(`[[...slug]]`) 안에 포함함으로써 선택적으로 만들 수 있습니다.

캐치 올 라우트와 선택적 캐치 올 라우트의 주요 차이점은, 선택적 캐치 올 라우트를 사용하면 매개변수 없는 라우트도 일치시킨다는 것입니다(`/api/post`가 위의 예시입니다.)

`query` 객체는 다음과 같이 표현됩니다.

```json
{ } // GET `/api/post` (빈 객체)
{ "slug": ["a"] } // `GET /api/post/a` (단일 요소 배열)
{ "slug": ["a", "b"] } // `GET /api/post/a/b` (다중 요소 배열)
```

### 주의사항

- 미리 정의된 API 라우트는 동적 API 라우트보다 우선시되고, 동적 API 라우트는 캐치 올 API 라우트보다 우선시됩니다. 다음 예제를 확인해 보세요.
  - `pages/api/post/create.js` - `/api/post/create`와 일치합니다.
  - `pages/api/post/[pid].js` - `/api/post/1`, `/api/post/abc` 등과 일치합니다. 하지만 `/api/post/create`와는 일치하지 않습니다.
  - `pages/api/post/[...slug].js` - `/api/post/1/2`, `/api/post/a/b/c` 등과 일치합니다. 하지만 `/api/post/create`, `/api/post/abc`와는 일치하지 않습니다

> ## 에지 API 라우트

Reviewer: ```suggestion
## 엣지 API 라우트
```
Edge를 엣지라고 많이 번역하는 것으로 아는데 한번 확인이 필요할 것 같습니다.

Author: edge를 `에지` 또는 `엣지` 둘 중에 어떤것을 사용해도 상관없을것 같습니다. 
다만, 웹 개발자에게 친숙한 angular, aws에서 edge를 엣지로 번역한 사례가 있기 때문에 `엣지`로 수정했습니다.

@chaejunlee 

- [angular에서 edge 번역](https://angular.kr/guide/service-worker-devops)
- [aws에서 edge 번역](https://aws.amazon.com/ko/what-is/edge-computing/)

}
```

## API 라우트와의 차이점

에지 API 라우트는 [에지 런타임](/docs/pages/api-reference/edge)을 사용하며, API 라우트는 [Node.js 런타임](/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes)을 사용합니다.

> 에지 API 라우트는 서버로부터 [응답을 스트림](/docs/pages/api-reference/edge#stream-apis)할 수 있으며, 캐시된 여러 파일(예: HTML, CSS, JavaScript)에 접근한 *이후*에 실행됩니다. 서버 측 스트리밍은 더 빠른 [Time To First Byte (TTFB)](https://web.dev/ttfb/)로 성능을 향상할 수 있습니다.

Reviewer: ```suggestion
에지 API 라우트는 서버로부터 [응답을 스트림](/docs/pages/api-reference/edge#stream-apis)할 수 있으며, 캐시된 여러 파일(예: HTML, CSS, JavaScript)에 접근한 _이후_에 실행됩니다. 서버 측 스트리밍은 더 빠른 [Time To First Byte (TTFB)](https://web.dev/ttfb/)로 성능을 향상할 수 있습니다.
```
